### ISSUE TITLE
read_xpm can't handle XPM files generated by ImageMagick and maybe other sources

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#4081]
Imported from SourceForge on 2024-07-09 11:17:39
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-11-26 22:19:19
Original: https://sourceforge.net/p/maxima/bugs/4218

---

`read_xpm` in the share package `draw` runs into an error when I try to load an XPM file created by ImageMagick. Looks like the .xpm file contains C comments which are unexpected by read_xpm.

The attached file `gcl1.xpm-original` is the unmodified ImageMagick output. The attached file `gcl1.xpm-modified` is the same except comments after the first are removed.

`read_xpm` successfully reads the modified file but not the original.

Not sure how to fix up `read_xpm`. Since C code is allowed, images are potentially differently arranged line by line. Given that, one approach is to use the Maxima parser to burn off the first several tokens (calling SCAN-ONE-TOKEN-G) and then parser the remaining array of strings, comments and all (calling MREAD). Possibly then rewrite the string array as XPM2 (non-C, simple data file) and then process that.

Attachments:
- https://sourceforge.net/p/maxima/bugs/4218/attachment/gcl1.xpm-modified
- https://sourceforge.net/p/maxima/bugs/4218/attachment/gcl1.xpm-original

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-09 11:17:41
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-12-06 22:13:25
Original: https://sourceforge.net/p/maxima/bugs/4218/#8b97

---

Yeah, I think the way to go is to modify take_channel, etc. to handle rgb + alpha in addition to just rgb. They are pretty simple functions, I think it will be easy.

I looked at libXpm and also ImageMagick which has functions to handle XPM files, and got some inspiration there. I'm not going to try to handle XPM1 or XPM2; maybe someone else will want to work on it.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-09 11:17:45
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-12-12 18:54:09
Original: https://sourceforge.net/p/maxima/bugs/4218/#0112

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-09 11:17:48
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-12-12 18:54:09
Original: https://sourceforge.net/p/maxima/bugs/4218/#8e65

---

I fixed the original problem by allowing the XPM parser to handle a wider range of inputs (maybe not all, however, I tried `read_xpm` on all the 100 or so XPM images on my Linux system and it works as expected for all of them). Also I fixed up the other functions for operations on `rgb` pictures to work with `rgb_alpha` as returned by `read_xpm`. 

For the record one can see the changes I made via `git log c13151d..5842e95 -- share/draw`.

Closing this issue as fixed. I'm still planning to update the documentation for the picture functions but that's beyond the scope of this ticket.
